add enhance logs and extract timestamp docs (#3118)
elastic/observability-docs: docs/en/observability/logs-stream.asciidoc

[[logs-stream]]
= Stream a log file

In this guide, you'll learn how to send a log file to Elasticsearch using a standalone {agent}, configure the {agent} and your data streams using the `elastic-agent.yml` file, and query your logs using the data streams you've set up. Once your log files are in {es}, see the <<logs-stream-enhance-logs>> section to learn how to parse your log data and extract fields so you can filter and sort your logs effectively.

[discrete]

> [[logs-stream-prereq]]

Reviewer: (continued from Slack)

Float (or discrete) these at L1 headings so that you can bump the heading level up for every topic below this

- Your API key is in `<id>:<key>` format. If not, your API key may be in an unsupported format, and you'll need to create an API key in *Beats* format. 

If you're still running into issues, see {fleet-guide}/fleet-troubleshooting.html[{agent} troubleshooting] and {fleet-guide}/elastic-agent-configuration.html[Configure standalone Elastic Agents].

[discrete]
[[logs-stream-enhance-logs]]
== Get the most out of your log data

Make your logs more useful by extracting structured fields from your unstructured log data. Extracting structured fields makes it easier to search, analyze, and filter your log data. 

Let's look at this log example:

[source,log]
----
2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.
----

Add this log to {es} using the following command in *Dev Tools*, found in your deployment in the left navigation under *Management*:

[source,console]
----
POST logs-example-default/_doc
{
  "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
}
----

Use the this command to look at the document's details:

[source,console]
----
GET /logs-example-default/_search
----

You see something like this:

[source,JSON]
----

> {
> ...
>       {
>         "_index": ".ds-logs-example-default-2023.08.09-000001",
>         "_id": "RsWy3IkB8yCtA5VGOKLf",
>         "_score": 1,
>         "_source": {
>           "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.",
>           "@timestamp": "2023-08-09T17:19:27.73312243Z"
>         }
>       }
>     ]
>   }
> }

Reviewer: I'm nitpicking here, but when shortening these, I think it's important to maintain the hierarchy as much as possible. The content we're interested in is `hits.hits[0].*`, so I'd shorten like this instead: 

```suggestion
{
  ...
  "hits": {
    ...
    "hits": [
      {
        "_index": ".ds-logs-example-default-2023.08.09-000001",
        "_id": "RsWy3IkB8yCtA5VGOKLf",
        "_score": 1,
        "_source": {
          "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.",
          "@timestamp": "2023-08-09T17:19:27.73312243Z"
        }
      }
    ]
  }
}
```

- `data_stream` – Enables data streams.
- `priority` – Index templates with higher priority take precedence over lower priority. If a data stream matches multiple index templates, the template with the higher priority is used. Built-in templates have a priority of `200`, so we recommend a priority higher than `200`.
- `index.default_pipeline` – The name of your ingest pipeline. `logs-example-default` in this case.

[discrete]
[[logs-stream-create-data-stream]]
==== Create your data stream

Create your data stream using the {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme]. The name needs to match the name of your pipeline. For this example, we'll name the data stream `logs-example-default` and use the example log:

[source,console]
----
POST logs-example-default/_doc
{
  "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
}
----

Now look at the details using this command:

[source,console]
----
GET /logs-example-default/_search
----

You can see the pipeline extracted `@timestamp` field:

[source,JSON]
----

> {
> ...
>       {
>         "_index": ".ds-logs-example-default-2023.08.09-000001",
>         "_id": "RsWy3IkB8yCtA5VGOKLf",
>         "_score": 1,
>         "_source": {
>           "message": "WARN 192.168.1.101 Disk usage exceeds 90%.",
>           "@timestamp": "2023-08-08T13:45:12.123Z"
>         }
>       }
>     ]
>   }
> }

Reviewer: See above comment

```suggestion
{
  ...
  "hits": {
    ...
    "hits": [
      {
        "_index": ".ds-logs-generic-default-2023.08.09-000001",
        "_index": ".ds-logs-example-default-2023.08.09-000001",
        "_id": "RsWy3IkB8yCtA5VGOKLf",
        "_score": 1,
        "_source": {
          "message": "WARN 192.168.1.101 Disk usage exceeds 90%.",
          "@timestamp": "2023-08-08T13:45:12.123Z"
        }
      }
      }
    ]
  }
}
```

- *@timestamp* – `2023-08-08T13:45:12.123Z` – Extracting this field lets you sort logs by date and time. This is helpful when you want to view your logs in the order that they occurred or identify when issues happened.
- *log.level* – `WARN` – Extracting this field lets you filter logs by severity. This is helpful if you want to  focus on high-severity WARN or ERROR-level logs, and reduce noise by filtering out low-severity INFO-level logs.
- *host.ip* – `192.168.1.101` – Extracting this field lets you filter logs by the host IP addresses. This is helpful if you want to focus on specific hosts that you’re having issues with or if you want to find disparities between hosts.
- *message* – `Disk usage exceeds 90%.`

NOTE: These fields are part of the {ecs-ref}/ecs-reference.html[Elastic Common Schema (ECS)]. The ECS defines a common set of fields that you can use across Elasticsearch when storing data, including log and metric data.

See these sections information on extracting specific fields:

> - <<logs-stream-extract-timestamp>>

Reviewer 2: Are we planning to add more sections to extract the other fields, such as `log.level` and `host.ip`?

Author: Yep, goal is to merge this shortly and continue to "Extract the `log.level`". This [meta issue](https://github.com/elastic/observability-docs/issues/3115) outlines the plan.

[discrete]
[[logs-stream-extract-timestamp]]
=== Extract the `@timestamp` field

This section shows you how to extract the `@timestamp` field from the example log:

[source,log]
----
2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.
----

To extract the timestamp you need to:

- <<logs-stream-ingest-pipeline>>
- <<logs-stream-simulate-api>>
- <<logs-stream-index-template>>
- <<logs-stream-create-data-stream>>

[discrete]
[[logs-stream-ingest-pipeline]]
==== Use an ingest pipeline to extract the `@timestamp`

To extract the `@timestamp` field, use an ingest pipeline with a dissect processor and a date processor. Processors are executed in the order they are defined. The example pipeline runs these processors: 

. The {ref}/dissect-processor.html[dissect processor] extracts a structured field from your unstructured log message based on a pattern you set. In the following example, the dissect processor extracts the timestamp to the `timestamp` field.

> . The {ref}/date-processor.html[date processor] then takes the `timestamp` field extracted by the dissect processor and extracts it to a date field. By default, this date field is called `@timestamp`. The date processor can also set the timezone, change the target field, and change the output format of the timestamp.

Reviewer 2: I don't think the date processor is required in this case. ES will be able to parse string timestamps. According to https://www.elastic.co/guide/en/elasticsearch/reference/current/date.html, the default format for `date` fields is `strict_date_optional_time||epoch_millis`. According to https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html, `strict_date_optional_time` supports the formats `yyyy-MM-dd'T'HH:mm:ss.SSSZ` or ` yyyy-MM-dd`. As the timestamp in the log file already uses that format, the date processor is not necessary.

However, if the date looks a bit more complex and is non-standard, the date processor is certainly necessary.

I think in this case, I'd remove the date processor but add a link to the date processor docs and explain that it can be useful for parsing dates that are not in ISO or format epoch_millis.

This command creates an ingest pipeline with a dissect and a date processor:

[source,console]
----
PUT _ingest/pipeline/logs-example-default
{
  "description": "Extracts the timestamp from log",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{@timestamp} %{message}"
      }
    },
    {
      "date": {
        "field": "@timestamp",
        "formats": ["ISO8601"]
      }
    }
  ]
}
----

. Set these values for the dissect processor:
- `field` – The field you're extracting data from, `message` in this case.
- `pattern`– The pattern of the elements in your log data:
+
[source,JSON]
----
%{timestamp} %{message}
----
+

> This extracts the timestamp, `2023-08-08T13:45:12.123Z`, to the `timestamp` field, and the rest of the message, `WARN 192.168.1.101 Disk usage exceeds 90%.`, stays in the `message` field.

Reviewer: and to while? idk

```suggestion
This extracts the timestamp, `2023-08-08T13:45:12.123Z`, to the `timestamp` field, while the rest of the message, `WARN 192.168.1.101 Disk usage exceeds 90%.`, stays in the `message` field.
```

+
. Set these values for the date processor:

> - `_ingest/pipeline/logs-example-default` – The name of the pipeline,`logs-example-default`, needs to match the name of your data stream. You'll set up your data stream in the next section. See {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme] for more information.

Reviewer: this link doesn't render

[source,console]
----
PUT _index_template/logs-example-default-template
{
  "index_patterns": [ "logs-example-default" ],
  "data_stream": { },
  "priority": 500,
  "template": {
    "settings": {
      "index.default_pipeline":"logs-example-default"
    }
  }
}
----

Set the following values for the index template:

- `index_patterns`– The index pattern needs to match your log data stream. Naming conventions for data streams are `<type>-<dataset>-<namespace>`. In this example, your logs data stream is named `logs-example-default`. Data that matches this pattern will go through your pipeline.
- `data_stream` – Enables data streams.
- `priority` – Index templates with higher priority take precedence over lower priority. If a data stream matches multiple index templates, the template with the higher priority is used. Built-in templates have a priority of `200`, so we recommend a priority higher than `200`.
- `index.default_pipeline` – The name of your ingest pipeline. `logs-example-default` in this case.

[discrete]
[[logs-stream-create-data-stream]]
==== Create your data stream

> Create your data stream using the {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme]. The name needs to match the name of your pipeline. For this example, we'll name the data stream `logs-example-default` and use the example log:

Reviewer: This link doesn't render

[source,console]
----
POST logs-example-default/_doc
{
  "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
}
----

> Now look at the details using this command:

Reviewer: Look at the details of what?

[source,console]
----
GET /logs-example-default/_search
----

You can see the pipeline extracted the `@timestamp` field:

[source,JSON]
----
{
...
{
  ...
  "hits": {
    ...
    "hits": [
      {

>         "_index": ".ds-logs-generic-default-2023.08.09-000001",

Reviewer 3: This seems off?

        "_index": ".ds-logs-example-default-2023.08.09-000001",
        "_id": "RsWy3IkB8yCtA5VGOKLf",
        "_score": 1,
        "_source": {
          "message": "WARN 192.168.1.101 Disk usage exceeds 90%.",
          "@timestamp": "2023-08-08T13:45:12.123Z"
        }
      }
    ]
  }
}
----

You can now use the `@timestamp` field to sort your logs by the date and time they happened.

[discrete]
[[logs-stream-timestamp-troubleshooting]]
=== Troubleshoot your `@timestamp` field

Check the following common issues for possible solutions:

> - *Timestamp failure* – If your data has inconsistent date formats, you can set `ignore_failure` to `true` for your date processor. This process logs with correctly formatted dates and ignore those with issues.

Reviewer: ```suggestion
- *Timestamp failure* – If your data has inconsistent date formats, you can set `ignore_failure` to `true` for your date processor. This processes logs with correctly formatted dates and ignores those with issues.
```

- *Incorrect timezone* – Set your timezone using the `timezone` option on the {ref}/date-processor.html[date processor].

> - *Incorrect timestamp format* – Your timestamp can be a java time pattern or one of the following formats: ISO8601, UNIX, UNIX_MS, or TAI64N. See the {ref}/mapping-date-format.html[mapping date format] for more information on timestamp formats.

Reviewer: ```suggestion
- *Incorrect timestamp format* – Your timestamp can be a Java time pattern or one of the following formats: ISO8601, UNIX, UNIX_MS, or TAI64N. See the {ref}/mapping-date-format.html[mapping date format] for more information on timestamp formats.
```

You can set additional fields with the date processor to update timezones, set a different output field, and change the output format of your timestamp. For more on this, see the {ref}/date-processor.html[date processor] documentation.

[discrete]
[[logs-stream-simulate-api]]
==== Test your pipeline with the simulate pipeline API

Make sure your pipeline is working as expected with the {ref}/simulate-pipeline-api.html#ingest-verbose-param[simulate pipeline API]. Run this command to test your pipeline:

[source,console]
----
POST _ingest/pipeline/logs-example-default/_simulate
{
  "docs": [
    {
      "_source": {
        "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
      }
    }
  ]
}
----

The results should show the `@timestamp` field extracted from the `message` field:

[source,console]
----
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_version": "-3",
        "_source": {
          "message": "WARN 192.168.1.101 Disk usage exceeds 90%.",
          "@timestamp": "2023-08-08T13:45:12.123Z"
        },
        "_ingest": {

>           "timestamp": "2023-08-10T20:37:29.090624429Z"

Reviewer 3: Was this timestamp automatically created based on the above pipeline?

Author: Yeah, this was the result I got when running the simulate pipeline api.

Reviewer 3: Interesting. Lets leave it as this. Will dig into this on my end.

        }
      }
    }
  ]
}
----

NOTE: Create the index pipeline using the `PUT` command in the previous section before using the simulate pipeline API.

[discrete]
[[logs-stream-index-template]]
==== Configure your data stream with an index template

After creating your ingest pipeline, create an index template to point your log data to your pipeline using this command:

[source,console]
----

> PUT _index_template/logs-example-default-template

Reviewer 3: @felixbarny @felixbarny If someone is building their own dataset, I assume they should not use `@custom` but the (future) `@default`?

Reviewer 4: IIUC, this example is about registering a custom index templates with higher priority to override the default one. In such cases, neither `@custom` nor `@default` should be used and the only recommendation is to avoid using `@`. If that's the case, this looks right as it is

Reviewer 3: What the user is building here is their custom dataset without having an instance of an integration. My assumption, a user could convert it to it eventually. 

As it is the user owned dataset, I'm tempted to recommend to use `@default`. It means, if someone else would want to extend the dataset could use `@custom`. 

> In such cases, neither @custom nor @default should be used and the only recommendation is to avoid using @. If that's the case, this looks right as it is

@eyalkoren In the context of the data stream naming scheme, should "random" templates ever be used? @joshdover How would Fleet keep track of these?

@weltenwort Also bringing you in here as I think this discussion could become interesting in the context of creating integrations.

@mdbirnstiehl Please don't block on this conversation, we can always update it.

Reviewer 4: > @eyalkoren In the context of the data stream naming scheme, should "random" templates ever be used?

Again, my understanding is that this depends on who owns/registers the templates. The way I see it, the naming scheme is only relevant for templates managed by Elastic (never mind which component). This frees us from keeping track of anything else. We make best efforts to prevent clashes **between custom components and such managed by Elastic** by documenting that `@` should be avoided, but we are not eligible for preventing clashes between custom components. I hope this makes sense.

Reviewer 2: In the future, I think that ideally users shouldn't need to create an index template but rather hook into extension points for the managed index templates. See also https://github.com/elastic/elasticsearch/issues/97664.

But for the time being, creating a custom `logs-example-default-template` is fine. The only thing I'd change here is to use the built-in index template for `logs-*-*` as a foundation for the custom index template. The built-in template can be retrieved via `GET _index_template/logs`.

Reviewer 5: > How would Fleet keep track of these?

We haven't designed how someone would do this conversion from a existing index template and ingest pipeline to a proper package, so I can only speculate at this point.

The way I'd imagine it working though is that we could extract any settings defined at the top level and move them into the `@package` settings later (strictly speaking, we'd put them the package-spec definition, in the data stream's `manifest.yml` `elasticsearch.settings` block)

Reviewer 3: I think there should be an option (eventually) to add a dataset to an integration. Not sure we can make this work if the dataset itself doesn't already follow the convention around naming.

@weltenwort Have you thought more about this?

Reviewer 6: I tend to agree with what @joshdover hinted at:

- Guidance for the "manual quick-start" approach avoids any names that might collide with integration-managed assets (i.e. avoid `@`s).
- The "uplift into an integration" process would pick settings from the user-created assets and use them to create proper integration-managed assets. If those are ambiguous or conflicting we could let the user choose during the workflow.

The advantage of this approach would be that we don't front-load all the naming knowledge, which could increase the hurdle to get data in. It would also mean the instructions wouldn't have to change frequently as our conventions evolve. We "just" need to adapt our conversion routine.

- `data_stream` – Enables data streams.
- `priority` – Index templates with higher priority take precedence over lower priority. If a data stream matches multiple index templates, the template with the higher priority is used. Built-in templates have a priority of `200`, so we recommend a priority higher than `200`.
- `index.default_pipeline` – The name of your ingest pipeline. `logs-example-default` in this case.

[discrete]
[[logs-stream-create-data-stream]]
==== Create your data stream

Create your data stream using the {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme]. The name needs to match the name of your pipeline. For this example, we'll name the data stream `logs-example-default` and use the example log:

[source,console]
----
POST logs-example-default/_doc
{
  "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
}
----

Now look at the details using this command:

[source,JSON]
----
GET /<index-name>/_doc/<document-id>
----

Under `_source`, you can see the `timestamp` field that was extracted by the dissect processor and the `@timestamp` field that was extracted by the date processor:

[source,JSON]
----
{

>   "took": 0,

Reviewer 3: For these kind of result documents, I suggest we filter out all the "non important" parts like `_shards` total hits etc. to make it focused on what matters in this case. Or do you think that is more confusing?

How can we indicate there is more content? For example using `...`

Reviewer: ++ we do this elsewhere in the docs too

Author: I was thinking about this as well. I'll see if we have any examples in our current docs to see if we have a consistent approach for this situation.

Author: I guess I was commenting on an outdated version of the PR.

If you're still running into issues, see {fleet-guide}/fleet-troubleshooting.html[{agent} troubleshooting] and {fleet-guide}/elastic-agent-configuration.html[Configure standalone Elastic Agents].

[discrete]
[[logs-stream-enhance-logs]]
== Get the most out of your log data

Now that you've sent your log data to {es}, this section helps you make sure you're getting the most out of it. You can do this by taking your unstructured log data and using ingest pipelines to extract structured fields. Extracting structured fields from your log data makes it easier to search, analyze, and filter your log data. 

Let's look at the following log data example:

[source,log]
----
2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.
----

This log is made up of the following potential fields:

- *timestamp*: "2023-08-08T13:45:12.123Z"
- *log.level*: "WARN"
- *host.ip*: "192.168.1.101"
- *message*: "Disk usage exceeds 90%."

Extracting these structured fields from your log data lets you search or filter your data in the following ways:

> - *timestamp* – Extracting this field lets you sort logs by date and time. This is helpful when you want to view your logs in the order that they occurred or identify when issues happened.

Reviewer 3: lets mention that these are also ECS fields which should be the same across all documents which allows correlation.

//add link for dissect processor

- the `field` you're extracting data from (`message` in this case).
- the `pattern` of your log data. To extract the timestamp from the example, use the following pattern:
+
[source,JSON]
----
%{@timestamp} %{message}
----
+
This puts the timestamp (2023-08-08T13:45:12.123Z) in the `@timestamp` field and the rest of the message (WARN 192.168.1.101 Disk usage exceeds 90%.) in the `message` field.

For the `date` processor, you need to set following values:
//add link for date processor

- The `field` you're extracting the date from (`@timestamp` in this case).
- The `formats` of the current timestamp (`ISO8601` in this case). The date processor also accepts Java time patterns or the following formats: `ISO8601`, `UNIX`, `UNIX_MS`, `TAI64N`.

Optionally, you can set the following values for teh `date` processor:
- The `timezone` to use when parsing the date.
- The `output_format` if you want to change the format when writing the date to the `@timestamp` field. Must be a valid java time pattern.
//need link for java time pattern

Define your ingest pipeline with the dissect and date processors with the following:

[source,JSON]
----

> PUT _ingest/pipeline/my-pipeline

Reviewer 3: Two things here

* Before we have an ingest pipeline, lets make sure we have first an example on ingesting the log message we had above. The data stream to put it in must match the data stream naming scheme like `logs-foo-bar`
* The second point is that the ingest pipeline names we use also now must match that index, which is important as otherwise things will not fit together.

{
  "description": "Pipeline to extract timestamp from log data",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "patterns": "%{@timestamp} %{message}"
      }
    },
    {
      "date": {
        "field": "@timestamp",
        "formats": ["current-timestamp-format"],
        "timezone": ["desired-timezone"],
        "output_format": ["desired-timezone-output-format"]
      }
    }
  ]
}
----

> 

Reviewer 3: I would now ingest the document again and show the user the result. Also show the user how to query the data / see the result and post it here. Anyone reading this should be able to copy paste the examples and be successful in going through the full flow. 

Eventually users can put in their own message and go through the same iterations.

If you're still running into issues, see {fleet-guide}/fleet-troubleshooting.html[{agent} troubleshooting] and {fleet-guide}/elastic-agent-configuration.html[Configure standalone Elastic Agents].

[discrete]
[[logs-stream-enhance-logs]]
== Get the most out of your log data

This section shows you how to get the most out of your log data after sending it to {es} by extracting structured fields from your unstructured log data. Extracting structured fields makes it easier to search, analyze, and filter your log data. 

Let's look at the following log example:

[source,log]
----
2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.
----

Add this log to {es} using the following command in *Dev Tools*, found in your deployment in the left navigation under *Management*:

[source,console]
----
POST logs-generic-default/_doc
{

>   "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."

Reviewer 3: Great to see this working. Before 8.10 this would have thrown an error.

@felixbarny For debugging purpose, it would be nice if there was an indication on the doc that it did it.

Author: I was testing on 8.9 and didn't get an error.

Reviewer 2: > @felixbarny For debugging purpose, it would be nice if there was an indication on the doc that it did it.

I don't disagree but I'm not sure whether adding additional storage for that additional piece of debugging metadata is worth it.

}
----

Use the following command to look at the document's details:

[source,console]
----

> GET /<index-name>/_doc/<document-id>

Author: @ruflin I'm not sure this is the best way to do this. Do you have a better suggestion here?

Reviewer 3: I would use the search command `GET /logs-generic-default/_search`. It will now only contain 1 doc and works even if the user does not know the doc id.

----

You see something like the following:

[source]
----
{
  "_index": ".ds-logs-generic-default-2023.08.09-000001",
  "_id": "sERP24kB4jRKLxIff9f1",
  "_version": 1,
  "_seq_no": 4,
  "_primary_term": 1,
  "found": true,
  "_source": {
    "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.",
    "@timestamp": "2023-08-09T17:19:27.73312243Z"
  }
}
----

Notice the text in your message isn't parsed, so you can't filter by any individual fields. Also, the `@timestamp` field shows when you added the data to {es}, not when the log occurred. In this format, neither of these fields are particularly useful for sorting or filtering. Your message, however, has all of these potential fields:

> - *timestamp* – 2023-08-08T13:45:12.123Z – Extracting this field lets you sort logs by date and time. This is helpful when you want to view your logs in the order that they occurred or identify when issues happened.

Reviewer 3: ```suggestion
- *@timestamp* – 2023-08-08T13:45:12.123Z – Extracting this field lets you sort logs by date and time. This is helpful when you want to view your logs in the order that they occurred or identify when issues happened.
```

[source,log]
----
2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.
----

To extract the timestamp you need to:
- <<logs-stream-ingest-pipeline>>
- <<logs-stream-index-template>>
- <<logs-stream-create-data-stream>>

[discrete]
[[logs-stream-ingest-pipeline]]
==== Extract the timestamp with an ingest pipeline

To extract the `@timestamp` field, use an ingest pipeline with a dissect processor and a date processor. Processors are executed in the order they are defined, so in the following example: 

. The {ref}/dissect-processor.html[dissect processor] extracts a structured field from your unstructured log message based on a pattern you set. In the following example, it extracts the timestamp to the `timestamp` field.
. The {ref}/date-processor.html[date processor] then takes the `timestamp` field extracted by the dissect processor and extracts it to a date field. By default, this date field is called `@timestamp`. The date processor can also set the timezone, change the target field, and change the output format of the timestamp.

The following example creates an ingest pipeline with a dissect and a date processor:

[source,console]
----

> PUT _ingest/pipeline/logs-generic-default

Reviewer 3: @eyalkoren We need to name this based on the upcoming conventions. Can you chime in?

        "field": "timestamp",
        "formats": ["ISO8601"]
      }
    }
  ]
}
----

. Set the following values for the dissect processor:
- `field` – The field you're extracting data from, `message` in this case.
- `pattern`– The pattern of the elements in your log data:
+
[source,JSON]
----
%{timestamp} %{message}
----
+
This extracts the timestamp, `2023-08-08T13:45:12.123Z`, to the `timestamp` field, and the rest of the message, `WARN 192.168.1.101 Disk usage exceeds 90%.`, stays in the `message` field.
+
. Set the following for the date processor:
- `_ingest/pipeline/logs-generic-default` – `logs-generic-default` is the name of the pipeline. The name of your pipeline needs to match the name of your data stream. You'll set up your data stream in the next section. See {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme] for more information.
- `field` – The field you're extracting the date from, `@timestamp` in this case.
- `formats` – The format of the current timestamp, `ISO8601` in this case. The date processor accepts {ref}/mapping-date-format.html[Java time patterns] or the following formats: `ISO8601`, `UNIX`, `UNIX_MS`, `TAI64N`.

> Optionally, you can set the following values for the `date` processor:

Reviewer 3: as this is not needed for the current example, I would link to the docs here instead.

- `timezone` – Sets the timezone to use when parsing the date.
- `target_field` – Sets the output field for the timestamp.
- `output_format` – Changes the format when writing the date to the timestamp field. Must be a valid java time pattern.

> [discrete]

Reviewer 3: before we create an index template, we can use the simulate API so the user sees direct results if it will work as expected and then hook up everything together: https://www.elastic.co/guide/en/elasticsearch/reference/current/simulate-pipeline-api.html#:~:text=You%20can%20use%20the%20simulate,verbose%20parameter%20to%20the%20request.

[[logs-stream-index-template]]
==== Configure your data stream with an index template

> After creating your ingest pipeline, you need to index your log data to a data stream using an index template. 

Reviewer 3: the data stream already uses a template, just not the one we have created but a default one.

- `index.default_pipeline` – The name of the ingest pipeline you created.

[discrete]
[[logs-stream-create-data-stream]]
==== Create your data stream

Create your data stream using the {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme]. The name needs to match the name of your pipeline. For this example, we'll use the default log data stream name, `logs-generic-default`, and the example log:

[source,console]
----
POST logs-generic-default/_doc
{
  "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
}
----

Now look at the details using the following command:

[source,JSON]
----
GET /<index-name>/_doc/<document-id>
----

Under `_source`, you can see the `timestamp` field that was extracted by the dissect processor and the `@timestamp` field that was extracted by the date processor:

[source]
----
{
  "_index": ".ds-logs-generic-default-2023.08.09-000001",
  "_id": "StG93IkByBSsmSqxzBqO",
  "_version": 1,
  "_seq_no": 3,
  "_primary_term": 1,
  "found": true,
  "_source": {
    "message": "WARN 192.168.1.101 Disk usage exceeds 90%.",

>     "timestamp": "2023-08-08T13:45:12.123Z",

Reviewer 3: why 2 tiemstamp fields?

Author: @ruflin I assumed this was a remnant of the dissect processor pulling a timestamp field and the date processor using that to pull the @timestamp field. Didn't realize that wasn't expected. I'll retest.

Author: I updated the dissect processor to use the `@timestamp` field instead of `timestamp` to correct this.

    "@timestamp": "2023-08-08T13:45:12.123Z"
  }
}
----

You can now use the `@timestamp` field to sort your logs by the date and time they happened.

[discrete]
[[logs-stream-timestamp-troubleshooting]]

> ==== Troubleshooting

Author: @ruflin do you have any suggestions for issues users might run into?

Reviewer 3: invalid timestamps, wrong time zone, figuring out the format. You have listed some of the config options above. But debugging it is painful.

@felixbarny @graphaelli I always found it very painful debugging "non standard" timestamp formats and trying to get things work. I wonder if we could improve this in the future too.

If you're still running into issues, see {fleet-guide}/fleet-troubleshooting.html[{agent} troubleshooting] and {fleet-guide}/elastic-agent-configuration.html[Configure standalone Elastic Agents].

[discrete]
[[logs-stream-enhance-logs]]
== Get the most out of your log data

> This section shows you how to get the most out of your log data after sending it to {es} by extracting structured fields from your unstructured log data. Extracting structured fields makes it easier to search, analyze, and filter your log data. 

Reviewer: > This section shows you how to get the most out of your log data after sending it to {es} by extracting structured fields from your unstructured log data.

I can't explain why, but I don't like this sentence. Here's a suggestion. You don't have to take it verbatim but maybe think more about how we open this section.

```
== Get the most out of your log data

Make your logs more useful by extracting structured fields from your unstructured log data.
Extracting structured fields makes it easier to search, analyze, and filter your log data. 
```

Let's look at the following log example:

[source,log]
----
2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.
----

Add this log to {es} using the following command in *Dev Tools*, found in your deployment in the left navigation under *Management*:

[source,console]
----
POST logs-example-default/_doc
{
  "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
}
----

Use the following command to look at the document's details:

[source,console]
----
GET /logs-example-default/_search
----

You see something like the following output:

> [source]

Reviewer: ```suggestion
[source,JSON]
```

----
{
  "took": 0,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": ".ds-logs-example-default-2023.08.09-000001",
        "_id": "RsWy3IkB8yCtA5VGOKLf",
        "_score": 1,
        "_source": {
          "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.",
          "@timestamp": "2023-08-09T17:19:27.73312243Z"
        }
      }
    ]
  }
}
----

Notice the text in your message isn't parsed, so you can't filter by any individual fields. Also, the `@timestamp` field shows when you added the data to {es}, not when the log occurred. In this format, neither of these fields are particularly useful for sorting or filtering. Your message, however, has all of these potential fields:

> - *@timestamp* – 2023-08-08T13:45:12.123Z – Extracting this field lets you sort logs by date and time. This is helpful when you want to view your logs in the order that they occurred or identify when issues happened.
> - *log.level* – WARN – Extracting this field lets you filter logs by severity. This is helpful if you want to  focus on high-severity WARN or ERROR-level logs, and reduce noise by filtering out low-severity INFO-level logs.
> - *host.ip* – 192.168.1.101 – Extracting this field lets you filter logs by the host IP addresses. This is helpful if you want to focus on specific hosts that you’re having issues with or if you want to find disparities between hosts.
> - *message* – Disk usage exceeds 90%.

Reviewer: ```suggestion
- *@timestamp* – `2023-08-08T13:45:12.123Z` – Extracting this field lets you sort logs by date and time. This is helpful when you want to view your logs in the order that they occurred or identify when issues happened.
- *log.level* – `WARN` – Extracting this field lets you filter logs by severity. This is helpful if you want to focus on high-severity WARN or ERROR-level logs, and reduce noise by filtering out low-severity INFO-level logs.
- *host.ip* – `192.168.1.101` – Extracting this field lets you filter logs by the host IP addresses. This is helpful if you want to focus on specific hosts that you’re having issues with or if you want to find disparities between hosts.
- *message* – `Disk usage exceeds 90%`.
```

> NOTE: These fields are part of the {ecs-ref}/ecs-reference.html[Elastic Common Schema (ECS)]. The ECS defines a common set of fields that you can use across Elasticsearch when storing data including log and metric data.

Reviewer: ```suggestion
NOTE: These fields are part of the {ecs-ref}/ecs-reference.html[Elastic Common Schema (ECS)]. The ECS defines a common set of fields that you can use across Elasticsearch when storing data, including log and metric data.
```

I'd add a comma here but up to you

> See the following sections for information on extracting specific fields:

Reviewer: ```suggestion
See the following sections for information on extracting specific fields:

```

Need a new line here to render the unordered list correctly

- <<logs-stream-extract-timestamp>>
- <<logs-stream-extract-loglevel>>
- <<logs-stream-extract-hostip>>

[discrete]
[[logs-stream-extract-timestamp]]
=== Extract the `@timestamp` field

This section shows you how to extract the `@timestamp` field from the following example log:

[source,log]
----
2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%.
----

> To extract the timestamp you need to:

Reviewer: ```suggestion
To extract the timestamp you need to:

```
Need a new line here to render the unordered list correctly

- <<logs-stream-ingest-pipeline>>
- <<logs-stream-index-template>>
- <<logs-stream-create-data-stream>>

[discrete]
[[logs-stream-ingest-pipeline]]
==== Use an ingest pipeline to extract the `@timestamp`

To extract the `@timestamp` field, use an ingest pipeline with a dissect processor and a date processor. Processors are executed in the order they are defined, so in the following example: 

> . The {ref}/dissect-processor.html[dissect processor] extracts a structured field from your unstructured log message based on a pattern you set. In the following example, it extracts the timestamp to the `timestamp` field.

Reviewer: I think we're saying "In the following example" too much here

. The {ref}/date-processor.html[date processor] then takes the `timestamp` field extracted by the dissect processor and extracts it to a date field. By default, this date field is called `@timestamp`. The date processor can also set the timezone, change the target field, and change the output format of the timestamp.

> The following example creates an ingest pipeline with a dissect and a date processor:

Reviewer: oop, here it is again

[source,console]
----
PUT _ingest/pipeline/logs-example-default
{
  "description": "Extracts the timestamp from log",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{@timestamp} %{message}"
      }
    },
    {
      "date": {
        "field": "@timestamp",
        "formats": ["ISO8601"]
      }
    }
  ]
}
----

. Set the following values for the dissect processor:
- `field` – The field you're extracting data from, `message` in this case.
- `pattern`– The pattern of the elements in your log data:
+
[source,JSON]
----
%{timestamp} %{message}
----
+
This extracts the timestamp, `2023-08-08T13:45:12.123Z`, to the `timestamp` field, and the rest of the message, `WARN 192.168.1.101 Disk usage exceeds 90%.`, stays in the `message` field.
+

> . Set the following for the date processor:

Reviewer: I'd change this to

```suggestion
. Set the following parameters for the date processor:
```

Reviewer: It looks like you say `Set the following values...` in an example below this. That works too. I'd update this to be consistent.

> - `_ingest/pipeline/logs-example-default` – `logs-example-default` is the name of the pipeline. The name of your pipeline needs to match the name of your data stream. You'll set up your data stream in the next section. See {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme] for more information.

Reviewer: With just a dash between two code snippets, it's a little tough to see what you're saying to do here. Not sure how to improve this though. Maybe it's fine

- `data_stream` – Enables data streams.
- `priority` – Index templates with higher priority take precedence over lower priority. If a data stream matches multiple index templates, the template with the higher priority is used. Built-in templates have a priority of `200`, so we recommend a priority higher than `200`.
- `index.default_pipeline` – The name of your ingest pipeline. `logs-example-default` in this case.

[discrete]
[[logs-stream-create-data-stream]]
==== Create your data stream

Create your data stream using the {fleet-ref}/data-streams.html#data-streams-naming-scheme[data stream naming scheme]. The name needs to match the name of your pipeline. For this example, we'll name the data stream `logs-example-default` and use the example log:

[source,console]
----
POST logs-example-default/_doc
{
  "message": "2023-08-08T13:45:12.123Z WARN 192.168.1.101 Disk usage exceeds 90%."
}
----

Now look at the details using the following command:

[source,JSON]
----
GET /<index-name>/_doc/<document-id>
----

Under `_source`, you can see the `timestamp` field that was extracted by the dissect processor and the `@timestamp` field that was extracted by the date processor:

> [source]

Reviewer: ```suggestion
[source,JSON]
```
